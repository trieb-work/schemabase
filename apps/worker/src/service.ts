import { Logger } from "@eci/util/logger";
import { IConsumer, Message } from "@eci/events-client";
import { Topic, EntryEvent } from "@eci/events/strapi";

export type EventSource<TTopic extends string, TMessage> = {
  consumer: IConsumer<TTopic, TMessage>;
  handlers: [{ topic: TTopic; handler: (message: TMessage) => Promise<void> }];
};

export type WorkerConfig = {
  logger: Logger;

  sources: {
    [name: string]: EventSource<Topic, Message<EntryEvent>>;
  };
};

export class Worker {
  private logger: Logger;
  private readonly sources: Record<
    string,
    EventSource<Topic, Message<EntryEvent>>
  >;

  constructor(config: WorkerConfig) {
    this.logger = config.logger;
    this.sources = config.sources;
  }

  start() {
    this.logger.info("Worker starting");
    for (const [name, source] of Object.entries(this.sources)) {
      for (const { topic, handler } of source.handlers) {
        source.consumer.consume(topic, handler);
        this.logger.info(`${name}: Started listening for ${topic} events`);
      }
    }
  }
}
