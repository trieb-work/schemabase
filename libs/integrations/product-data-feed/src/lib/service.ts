import ObjectsToCsv from "objects-to-csv";
import { htmlToText } from "html-to-text";
import { Product as RawProduct } from "@eci/adapters/saleor";
// @ts-expect-error it doesn't detec types for some reason
import edjsHTML from "editorjs-html";
import { generateUnitPrice } from "./generate-unit-price";
import { FeedVariant, Product } from "./types";

export interface ProductDataFeedService {
  generateCSV: (
    storefrontProductUrl: string,
    feedVariant: FeedVariant,
    channelSlug: string,
  ) => Promise<string>;
}

export type ProductDataFeedServiceConfig = {
  saleorClient: {
    getProducts: (variables: {
      first: number;
      channel: string;
    }) => Promise<RawProduct[]>;
  };
  channelSlug: string;
};

/**
 * Generate product data as .csv
 */

export class ProductDataFeedGenerator implements ProductDataFeedService {
  public readonly saleorClient: {
    getProducts: (variables: {
      first: number;
      channel: string;
    }) => Promise<RawProduct[]>;
  };
  public readonly channelSlug: string;

  public constructor(config: ProductDataFeedServiceConfig) {
    this.saleorClient = config.saleorClient;
    this.channelSlug = config.channelSlug;
  }

  public async generateCSV(
    storefrontProductUrl: string,
    feedVariant: FeedVariant,
  ): Promise<string> {
    const products = await this.generate(storefrontProductUrl, feedVariant);
    const csv = new ObjectsToCsv(products);
    return await csv.toString();
  }

  private async generate(
    storefrontProductUrl: string,
    feedVariant: FeedVariant,
  ): Promise<Product[]> {
    const rawProducts = await this.saleorClient.getProducts({
      first: 100,
      channel: this.channelSlug,
    });
    const products: Product[] = [];

    for (const rawProduct of rawProducts) {
      // we get the brand from a product attribute called brand
      const brand = rawProduct.attributes.find(
        (x) => x.attribute.name === "brand",
      )?.values[0]?.name;
      const googleProductCategory = rawProduct.attributes.find(
        (x) => x.attribute.name === "googleProductCategory",
      )?.values[0]?.name;

      // if we want to prefer the title instead of the seoTitle
      // const title = product.name ? product.name : product.seoTitle;

      const title = rawProduct.seoTitle ? rawProduct.seoTitle : rawProduct.name;

      let description = "";
      try {
        description = rawProduct.descriptionJson
          ? edjsHTML().parse(JSON.parse(rawProduct.descriptionJson))
          : rawProduct.seoDescription;
      } catch (err) {
        // console.warn(err)
      }

      description =
        feedVariant == "facebookcommerce"
          ? htmlToText(description)
          : description;

      const { hasVariants } = rawProduct.productType;

      if (!rawProduct.variants) {
        continue;
      }

      for (const variant of rawProduct.variants) {
        if (!variant) {
          continue;
        }

        const gtin = hasVariants
          ? variant.metadata?.find((x) => x?.key === "EAN")?.value
          : rawProduct.metadata?.find((x) => x?.key === "EAN")?.value;
        const unit_pricing_measure =
          variant.weight && rawProduct.weight
            ? generateUnitPrice(variant.weight, rawProduct.weight)
            : undefined;
        const product: Product = {
          id: variant.sku,
          title: hasVariants ? `${title} (${variant.name})` : title,
          description,
          rich_text_description:
            feedVariant === "facebookcommerce" ? description : undefined,
          image_link: hasVariants
            ? variant.images && variant.images.length > 0
              ? variant.images[0]?.url
              : ""
            : rawProduct.images && rawProduct.images.length > 0
            ? rawProduct.images[1]?.url
            : "",
          additional_image_link: hasVariants
            ? variant.images?.[1]?.url
            : rawProduct.images?.[2]?.url,
          link: storefrontProductUrl + rawProduct.slug,
          price: `${variant?.pricing?.priceUndiscounted?.gross.amount} ${variant?.pricing?.priceUndiscounted?.gross.currency}`,
          sale_price: `${variant?.pricing?.price?.gross.amount} ${variant.pricing?.price?.gross.currency}`,
          condition: "new",
          gtin,
          brand: brand ?? undefined,
          unit_pricing_measure,
          availability:
            variant.quantityAvailable < 1 || !rawProduct.isAvailableForPurchase
              ? "out of stock"
              : "in stock",
          google_product_category: googleProductCategory ?? undefined,
        };

        products.push(product);
      }
    }
    return products;
  }
}
