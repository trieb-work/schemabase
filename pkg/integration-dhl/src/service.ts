import { CronStateHandler } from "@eci/pkg/cronstate";
import {
  BullMQProducer,
  EventSchemaRegistry,
  Message,
  Topic,
} from "@eci/pkg/events";
import { id } from "@eci/pkg/ids";
import { ILogger } from "@eci/pkg/logger";
import { sleep } from "@eci/pkg/miscHelper/time";
import { DHLTrackingApp, PackageState, PrismaClient } from "@eci/pkg/prisma";
import { subMonths } from "date-fns";
import {
  Configuration,
  DefaultApi,
  SupermodelIoLogisticsTrackingShipmentEventStatusCodeEnum,
} from "./typescript-axios-client";

interface DHLTrackingSyncServiceConfig {
  dhlTrackingApp: DHLTrackingApp;
  db: PrismaClient;
  logger: ILogger;
}

export class DHLTrackingSyncService {
  private readonly logger: ILogger;

  public readonly dhlTrackingApp: DHLTrackingApp;

  private readonly db: PrismaClient;

  private readonly cronState: CronStateHandler;

  public constructor(config: DHLTrackingSyncServiceConfig) {
    this.logger = config.logger;
    this.dhlTrackingApp = config.dhlTrackingApp;
    this.db = config.db;
    this.cronState = new CronStateHandler({
      tenantId: this.dhlTrackingApp.tenantId,
      appId: this.dhlTrackingApp.id,
      db: this.db,
      syncEntity: "packageState",
    });
  }

  parseState = (
    state: SupermodelIoLogisticsTrackingShipmentEventStatusCodeEnum,
  ): PackageState | null => {
    switch (state) {
      case "pre-transit":
        return PackageState.INFORMATION_RECEIVED;

      case "transit":
        return PackageState.IN_TRANSIT;

      case "delivered":
        return PackageState.DELIVERED;

      case "failure":
        return PackageState.EXCEPTION;

      default:
        return null;
    }
  };

  private createAPIClient(apiKey: string): DefaultApi {
    const conf = new Configuration({ apiKey });
    return new DefaultApi(conf, "https://api-eu.dhl.com/track");
  }

  public async syncToECI(): Promise<void> {
    await this.cronState.get();
    const dhlClient = this.createAPIClient(this.dhlTrackingApp.apiKey);

    /// get all DHL packages, that are not delivered
    // with last status update older than 2 hours, to prevent too many API calls

    const dhlPackages = await this.db.package.findMany({
      where: {
        tenantId: this.dhlTrackingApp.tenantId,
        carrier: "DHL",
        state: {
          not: "DELIVERED",
        },
        trackingId: {
          not: null,
        },
        createdAt: {
          gt: subMonths(new Date(), 2),
        },
      },
    });

    this.logger.info(
      `Receiving ${dhlPackages.length} DHL packages, that we pull status updates from`,
    );

    for (const p of dhlPackages) {
      if (!p.trackingId) continue;
      this.logger.info(`Pulling package data from DHL for ${p.trackingId}`);

      const fullPackage = await dhlClient
        .shipmentsGet(
          p.trackingId,
          undefined,
          undefined,
          undefined,
          undefined,
          "en",
        )
        .catch((e) => {
          if ((e as any)?.response?.data?.status === 404) {
            this.logger.info(
              `No package with tracking Number ${p.trackingId} returned from DHL!`,
            );
          } else if ((e as any)?.response?.data?.status === 429) {
            this.logger.error(`Throttling Error from DHL: 429`);
          } else {
            this.logger.error(`Error pulling data from DHL:`, e);
          }
        });

      const shipment = fullPackage?.data?.shipments?.[0];

      if (!shipment) {
        continue;
      }
      if (!shipment.status?.statusCode || !shipment?.status?.timestamp)
        continue;
      const internalState = this.parseState(shipment.status?.statusCode);
      if (!internalState) {
        this.logger.error(
          `Could not parse package state ${shipment.status?.statusCode}` +
            `to our internal package state for ${p.trackingId}`,
        );
        continue;
      }
      this.logger.info(internalState);

      if (!this.dhlTrackingApp.trackingIntegrationId) {
        this.logger.info(
          `There is no tracking integration configured for DHL App ${this.dhlTrackingApp.id}.` +
            "Not updating package state",
        );
        continue;
      }

      const time = new Date(shipment.status.timestamp as string);
      const packageEvent: EventSchemaRegistry.PackageUpdate["message"] = {
        trackingId: p.trackingId,
        time: time.getTime() / 1000,
        location: shipment.status.location?.address?.addressLocality || "",
        state: internalState,
        trackingIntegrationId: this.dhlTrackingApp.trackingIntegrationId,
      };

      const bullmq = await BullMQProducer.new<
        EventSchemaRegistry.PackageUpdate["message"]
      >({
        topic: Topic.PACKAGE_UPDATE,
        tenantId: this.dhlTrackingApp.tenantId,
      });

      const message = new Message({
        header: {
          traceId: id.id("trace"),
        },
        content: packageEvent,
      });

      const { messageId } = await bullmq.produce(Topic.PACKAGE_UPDATE, message);
      this.logger.info(`Created bullmq message with ID ${messageId}`);
      await sleep(5000);
    }

    await this.cronState.set({
      lastRun: new Date(),
      lastRunStatus: "success",
    });
  }
}
