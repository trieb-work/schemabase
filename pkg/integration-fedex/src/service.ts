import { CronStateHandler } from "@eci/pkg/cronstate";
import {
    BullMQProducer,
    EventSchemaRegistry,
    Message,
    Topic,
} from "@eci/pkg/events";
import { id } from "@eci/pkg/ids";
import { ILogger } from "@eci/pkg/logger";
import { sleep } from "@eci/pkg/utils/time";
import { FedexTrackingApp, PackageState, PrismaClient } from "@eci/pkg/prisma";
import { subMonths } from "date-fns";
import axios, { AxiosInstance } from "axios";

interface FedexTrackingSyncServiceConfig {
    fedexTrackingApp: FedexTrackingApp;
    db: PrismaClient;
    logger: ILogger;
    testMode?: boolean;
}

type FedexTrackingPackage = {
    trackingNumber: string;
    trackResults: FedexTrackingResult[];
};

type FedexTrackingResult = {
    trackingNumberInfo: {
        trackingNumber: string;
        trackingNumberUniqueId: string;
        carrierCode: string;
    };
    additionalTrackingInfo: {
        nickname: string;
        packageIdentifiers: FedexPackageIdentifier[];
        hasAssociatedShipments: boolean;
    };
    shipperInformation: {
        contact: any; // TODO: Define the type for contact
        address: FedexAddress;
    };
    recipientInformation: {
        contact: any; // TODO: Define the type for contact
        address: FedexAddress;
    };
    latestStatusDetail: FedexStatusDetail;
    dateAndTimes: FedexDateTime[];
    availableImages: any[]; // TODO: Define the type for availableImages
    packageDetails: FedexPackageDetails;
    shipmentDetails: {
        possessionStatus: boolean;
        weight: FedexWeight[];
    };
    scanEvents: FedexScanEvent[];
    availableNotifications: string[];
    deliveryDetails: {
        actualDeliveryAddress: FedexAddress;
        deliveryAttempts: string;
        receivedByName: string;
        deliveryOptionEligibilityDetails: FedexDeliveryOptionEligibility[];
    };
    originLocation: {
        locationContactAndAddress: {
            address: FedexAddress;
        };
    };
    lastUpdatedDestinationAddress: FedexAddress;
    serviceDetail: {
        type: string;
        description: string;
        shortDescription: string;
    };
    standardTransitTimeWindow: {
        window: {
            ends: string;
        };
    };
    estimatedDeliveryTimeWindow: {
        window: any; // TODO: Define the type for estimatedDeliveryTimeWindow
    };
    goodsClassificationCode: string;
    returnDetail: any; // TODO: Define the type for returnDetail
};

type FedexPackageIdentifier = {
    type: string;
    values: string[];
    trackingNumberUniqueId: string;
    carrierCode: string;
};

type FedexAddress = {
    city: string;
    stateOrProvinceCode: string;
    countryCode: string;
    residential: boolean;
    countryName: string;
    streetLines?: string[];
    postalCode?: string;
};
const FedExTrackingCodes = {
    AA: "At Airport",
    AC: "At Canada Post facility",
    AD: "At Delivery",
    AF: "At FedEx Facility",
    AO: "Shipment arriving On-time",
    AP: "At Pickup",
    AR: "Arrived",
    AX: "At USPS facility",
    CA: "Carrier", // Given the conflict, chose the last occurrence
    CH: "Location Changed",
    DD: "Delivery Delay",
    DE: "Delivery Exception",
    DL: "Delivered",
    DP: "Departed",
    DR: "Vehicle furnished but not used",
    DS: "Vehicle Dispatched",
    DY: "Delay",
    EA: "Export Approved", // Given the conflict, chose the last occurrence
    ED: "Enroute to Delivery",
    EO: "Enroute to Origin Airport",
    EP: "Enroute to Pickup",
    FD: "At FedEx Destination",
    HL: "Hold at Location",
    IT: "In Transit",
    IX: "In transit (see Details)",
    LO: "Left Origin",
    OC: "Order Created",
    OD: "Out for Delivery",
    OF: "At FedEx origin facility",
    OX: "Shipment information sent to USPS",
    PD: "Pickup Delay",
    PL: "Plane Landed",
    PM: "In Progress",
    PU: "Picked Up",
    PX: "Picked up (see Details)",
    RR: "CDO requested",
    RM: "CDO Modified",
    RC: "Recipient", // Given the conflict, chose the last occurrence
    RS: "Return to Shipper",
    RP: "Return label link emailed to return sender",
    LP: "Return label link cancelled by shipment originator",
    RG: "Return label link expiring soon",
    RD: "Return label link expired",
    SE: "Shipment Exception",
    SF: "At Sort Facility",
    SP: "Split Status", // Last valid occurrence used
    TR: "Transfer",
    CC: "Cleared Customs",
    CD: "Clearance Delay",
    CP: "Clearance in Progress",
    SH: "Shipper",
    CU: "Customs",
    BR: "Broker",
    TP: "Transfer Partner",
} as const;

type FedExTrackingStatus = keyof typeof FedExTrackingCodes;

type FedexStatusDetail = {
    code: FedExTrackingStatus;
    derivedCode: string;
    statusByLocale: string;
    description: string;
    scanLocation: FedexAddress;
    ancillaryDetails: FedexAncillaryDetail[];
};

type FedexAncillaryDetail = {
    reason: string;
    reasonDescription: string;
    action: string;
    actionDescription: string;
};

type FedexDateTime = {
    type: string;
    dateTime: string;
};

type FedexWeight = {
    value: string;
    unit: string;
};

type FedexPackageDetails = {
    packagingDescription: {
        type: string;
        description: string;
    };
    physicalPackagingType: string;
    sequenceNumber: string;
    count: string;
    weightAndDimensions: {
        weight: FedexWeight[];
        dimensions: FedexDimension[];
    };
    packageContent: any[]; // TODO: Define the type for packageContent
};

type FedexDimension = {
    length: number;
    width: number;
    height: number;
    units: string;
};

type FedexScanEvent = {
    date: string;
    eventType: string;
    eventDescription: string;
    exceptionCode: string;
    exceptionDescription: string;
    scanLocation: FedexAddress;
    locationType: string;
    derivedStatusCode: string;
    derivedStatus: string;
};

type FedexDeliveryOptionEligibility = {
    option: string;
    eligibility: string;
};

type FedexResponse = {
    transactionId: string;
    output: {
        completeTrackResults: FedexTrackingPackage[];
    };
};

export class FedexTrackingSyncService {
    private readonly logger: ILogger;

    public readonly fedexTrackingApp: FedexTrackingApp;

    private readonly db: PrismaClient;

    private readonly cronState: CronStateHandler;

    public constructor(config: FedexTrackingSyncServiceConfig) {
        this.logger = config.logger;
        this.fedexTrackingApp = config.fedexTrackingApp;
        this.db = config.db;
        this.cronState = new CronStateHandler({
            tenantId: this.fedexTrackingApp.tenantId,
            appId: this.fedexTrackingApp.id,
            db: this.db,
            syncEntity: "packageState",
        });
    }

    /**
     * parse the fedex package status codes to our internal package state
     * @param state
     * @returns
     */
    parseState = (state: FedExTrackingStatus): PackageState | null => {
        switch (state) {
            case "PU":
            case "DL":
                return PackageState.DELIVERED;
            case "AD":
            case "OD":
                return PackageState.OUT_FOR_DELIVERY;
            case "DE":
                return PackageState.EXCEPTION;
            case "DD":
            case "PM":
            case "IT":
            case "DP":
            case "IX":
                return PackageState.IN_TRANSIT;
            case "OC":
                return PackageState.INFORMATION_RECEIVED;
            default:
                return null;
        }
    };

    private async getFedexPackage(
        trackingId: string,
        fedexClient: AxiosInstance,
    ) {
        try {
            const response = await fedexClient.post<FedexResponse>(
                "/track/v1/trackingnumbers",
                {
                    includeDetailedScans: true,
                    trackingInfo: [
                        {
                            shipDateBegin: "",
                            shipDateEnd: "",
                            trackingNumberInfo: {
                                trackingNumber: trackingId,
                                carrierCode: "FDXE",
                                trackingNumberUniqueId: "",
                            },
                        },
                    ],
                },
            );

            return response.data.output.completeTrackResults[0];
        } catch (e) {
            if (axios.isAxiosError(e)) {
                const error = e.response?.data;
                this.logger.error(
                    `Could not fetch package data from Fedex for ${trackingId}: ${JSON.stringify(
                        error,
                    )}`,
                );
            } else {
                this.logger.error(
                    `Could not fetch package data from Fedex for ${trackingId}: ${JSON.stringify(
                        e,
                    )}`,
                );
            }
            return null;
        }
    }

    private async createAPIClient(clientId: string, clientSecret: string) {
        // Oauth2 client credentials flow for Fedex. Token URL: https://apis.fedex.com/oauth/token
        // return a axios instance with the Bearer Token from the oauth2 flow set.
        // Send client credentials in body
        const token = await axios({
            method: "post",
            url: "https://apis.fedex.com/oauth/token",
            headers: {
                "Content-Type": "application/x-www-form-urlencoded",
            },
            data: `grant_type=client_credentials&client_id=${clientId}&client_secret=${clientSecret}`,
        });

        const instance = axios.create({
            baseURL: "https://apis.fedex.com",
            headers: {
                Authorization: `Bearer ${token.data.access_token}`,
                transId: id.id("trace"),
                transactionSrc: "schemabase",
            },
        });

        return instance;
    }

    public async syncToECI(testTrackingId?: string): Promise<void> {
        await this.cronState.get();
        const fedexClient = await this.createAPIClient(
            this.fedexTrackingApp.clientId,
            this.fedexTrackingApp.clientSecret,
        );

        /// get all Fedex packages, that are not delivered
        // with last status update older than 2 hours, to prevent too many API calls
        const fedexPackages = testTrackingId
            ? await this.db.package.findMany({
                  where: {
                      trackingId: testTrackingId,
                  },
              })
            : await this.db.package.findMany({
                  where: {
                      tenantId: this.fedexTrackingApp.tenantId,
                      carrier: "FEDEX",
                      state: {
                          not: "DELIVERED",
                      },
                      trackingId: {
                          not: null,
                      },
                      createdAt: {
                          gt: subMonths(new Date(), 2),
                      },
                      isTrackingEnabled: true,
                  },
              });

        this.logger.info(
            `Receiving ${fedexPackages.length} Fedex packages, that we pull status updates from`,
        );

        for (const p of fedexPackages) {
            if (!p.trackingId) continue;
            this.logger.info(
                `Pulling package data from Fedex for ${p.trackingId}`,
            );

            const fullPackage = await this.getFedexPackage(
                p.trackingId,
                fedexClient,
            );

            if (!fullPackage) {
                this.logger.error(
                    `Could not fetch package data from Fedex for ${p.trackingId}`,
                );
                continue;
            }

            const lastState = fullPackage.trackResults?.[0].latestStatusDetail;
            if (!lastState || !lastState.code) {
                this.logger.error(
                    `Package state from Fedex not including current status code: ${p.trackingId}`,
                    {
                        lastState,
                        trackingId: p.trackingId,
                    },
                );
                continue;
            }
            const internalState = this.parseState(lastState.code);
            if (!internalState) {
                this.logger.error(
                    `Could not parse package state ${JSON.stringify(
                        lastState,
                    )}` + `to our internal package state for ${p.trackingId}`,
                );
                continue;
            }

            /**
             * The status message coming from Fedex - like: "Processing at Fedex Facility"
             */
            const statusMessage =
                lastState.ancillaryDetails?.[0]?.reasonDescription ||
                lastState.description;

            // eslint-disable-next-line max-len
            const shipmentLocation = `${lastState.scanLocation.city}, ${lastState.scanLocation.stateOrProvinceCode}, ${lastState.scanLocation.countryCode}`;

            if (!this.fedexTrackingApp.trackingIntegrationId) {
                this.logger.info(
                    `There is no tracking integration configured for Fedex App ${this.fedexTrackingApp.id}.` +
                        "Not updating package state",
                );
                continue;
            }

            const lastScanEvent = fullPackage.trackResults[0].scanEvents[0];

            /**
             * Parse date & time
             */
            const time = new Date(lastScanEvent.date);
            const packageEvent: EventSchemaRegistry.PackageUpdate["message"] = {
                trackingId: p.trackingId,
                packageId: p.id,
                time: time.getTime() / 1000,
                location: shipmentLocation,
                state: internalState,
                trackingIntegrationId:
                    this.fedexTrackingApp.trackingIntegrationId,
                message: statusMessage,
            };

            const queue = await BullMQProducer.new<
                EventSchemaRegistry.PackageUpdate["message"]
            >({
                topic: Topic.PACKAGE_UPDATE,
                tenantId: this.fedexTrackingApp.tenantId,
            });

            const message = new Message({
                header: {
                    traceId: id.id("trace"),
                },
                content: packageEvent,
            });

            const { messageId } = await queue.produce(
                Topic.PACKAGE_UPDATE,
                message,
            );
            this.logger.info(`Created BullMQ message with ID ${messageId}`);

            await sleep(5000);
        }

        await this.cronState.set({
            lastRun: new Date(),
            lastRunStatus: "success",
        });
    }
}
