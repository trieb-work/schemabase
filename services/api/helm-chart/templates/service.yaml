apiVersion: v1
kind: Service
metadata:
  name: {{ .Release.Name }}
spec:
  clusterIP: None
  ports:
    - name: web
      port: 3000
      protocol: TCP
      targetPort: 3000
  selector:
    {{- include "schemabase-api.selectorLabels" . | nindent 4 }}
  sessionAffinity: None
  type: ClusterIP