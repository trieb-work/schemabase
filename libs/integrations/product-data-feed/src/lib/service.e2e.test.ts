import { ProductDataFeedGenerator } from "./service";
import { createGraphqlClient } from "@eci/graphql-client";

describe("e2e", () => {
  const generator = new ProductDataFeedGenerator({
    saleorGraphqlClient: createGraphqlClient(
      "https://pundf-test-api.triebwork.com/graphql/",
    ),
    channelSlug: "storefront",
  });

  it("returns all products in csv format", async () => {
    const csv = await generator.generateCSV("abc", "facebookcommerce");

    expect(csv).toMatchSnapshot();
  });
});
