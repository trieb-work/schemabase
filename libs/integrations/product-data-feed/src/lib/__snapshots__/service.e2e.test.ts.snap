// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`e2e returns all products in csv format 1`] = `
"id,title,description,rich_text_description,image_link,additional_image_link,link,price,sale_price,condition,gtin,brand,unit_pricing_measure,availability,google_product_category
pf-leb-3-gemischt,3 Lebkuchen in Bio Folie (),\\"xcbnxfn

,

ergh

,

ergh

,


H1\\",\\"xcbnxfn

,

ergh

,

ergh

,


H1\\",,,abc3-lebkuchen-in-bio-folie,5 EUR,5 EUR,new,,Pfeffer & Frost,,in stock,
pf-leb-5-gemischt,asdf (Mixed),Laridari,Laridari,,,abc5-lebkuchen-in-bio-folie,14.9 EUR,14.9 EUR,new,,Pfeffer & Frost,,in stock,
pf-dose-5-cd-dunkle,Designerdose Charlotte Dumortier (Dark Chocolate),,,https://s3.triebwork.com/static/products/pfefferundfrost_lebkuchen_charlotte_dunkle_schokolade_01_de.jpg,https://s3.triebwork.com/static/products/pfefferundfrost_lebkuchen_charlotte_dark_chocolate_01_en.jpg,abcdesignerdose-charlotte-dumortier,21.9 EUR,21.9 EUR,new,,,,in stock,
pf-dose-5-cd-gemischt,Designerdose Charlotte Dumortier (Mixed),,,https://s3.triebwork.com/static/products/pfefferundfrost_lebkuchen_charlotte_mix_01_de.jpg,https://s3.triebwork.com/static/products/pfefferundfrost_lebkuchen_charlotte_mixed_01_en.jpg,abcdesignerdose-charlotte-dumortier,21.9 EUR,21.9 EUR,new,,,,in stock,
pf-dose-5-cd-natur,Designerdose Charlotte Dumortier (Natural),,,https://s3.triebwork.com/static/products/pfefferundfrost_lebkuchen_charlotte_natural_elisen_01_en.jpg,https://s3.triebwork.com/static/products/pfefferundfrost_lebkuchen_charlotte_elisen_01_de.jpg,abcdesignerdose-charlotte-dumortier,21.9 EUR,21.9 EUR,new,,,,in stock,
pf-dose-5-cd-weis,Designerdose Charlotte Dumortier (White Chocolate),,,https://s3.triebwork.com/static/products/pfefferundfrost_lebkuchen_charlotte_weisse_schokolade_01_de.jpg,https://s3.triebwork.com/static/products/pfefferundfrost_lebkuchen_charlotte_white_chocolate_01_en.jpg,abcdesignerdose-charlotte-dumortier,21.9 EUR,21.9 EUR,new,,,,out of stock,
pf-dose-5-cd-zucker,Designerdose Charlotte Dumortier (Icing),,,https://s3.triebwork.com/static/products/pfefferundfrost_lebkuchen_charlotte_zuckerguss_01_de.jpg,https://s3.triebwork.com/static/products/pfefferundfrost_lebkuchen_charlotte_icing_01_en.jpg,abcdesignerdose-charlotte-dumortier,21.9 EUR,21.9 EUR,new,,,,in stock,
pf-dose-5-cj-dunkle,Lebkuchendose Edition Cachetejack (Dark Chocolate),,,https://s3.triebwork.com/static/products/pfefferundfrost_lebkuchen_dunkleschokolade_01_de.jpg,https://s3.triebwork.com/static/products/pfefferundfrost_lebkuchen_dunkleschokolade_01_en.jpg,abcdose-cachetejack,21.9 EUR,21.9 EUR,new,,,,in stock,
pf-dose-5-cj-gemischt,Lebkuchendose Edition Cachetejack (Mixed),,,https://s3.triebwork.com/static/products/pfefferundfrost_lebkuchen_mix_01_de.jpg,https://s3.triebwork.com/static/products/pfefferundfrost_lebkuchen_mix_01_en.jpg,abcdose-cachetejack,21.9 EUR,21.9 EUR,new,,,,in stock,
pf-dose-5-cj-natur,Lebkuchendose Edition Cachetejack (Natural),,,https://s3.triebwork.com/static/products/pfefferundfrost_elisenlebkuchen_01_de.jpg,https://s3.triebwork.com/static/products/pfefferundfrost_elisenlebkuchen_01_en.jpg,abcdose-cachetejack,21.9 EUR,21.9 EUR,new,,,,out of stock,
pf-dose-5-cj-weis,Lebkuchendose Edition Cachetejack (White Chocolate),,,https://s3.triebwork.com/static/products/pfefferundfrost_lebkuchen_wei%C3%9Feschokolade_01_de.jpg,https://s3.triebwork.com/static/products/pfefferundfrost_lebkuchen_wei%C3%9Feschokolade_01_en.jpg,abcdose-cachetejack,21.9 EUR,21.9 EUR,new,,,,out of stock,
pf-dose-5-cj-zucker,Lebkuchendose Edition Cachetejack (Icing),,,https://s3.triebwork.com/static/products/pfefferundfrost_lebkuchen_zuckerguss_01_de.jpg,https://s3.triebwork.com/static/products/pfefferundfrost_lebkuchen_zuckerguss_01_en.jpg,abcdose-cachetejack,21.9 EUR,21.9 EUR,new,,,,out of stock,
PFE-Gutschein-10,Geschenkgutscheine 10€ - 100€ (10€),\\"Der Pfeffer & Frost Geschenk Gutschein zum Einlösen im Shop - mache anderen eine
Freude

,

Bitte beachte: Es erfolgt KEIN Versand per Post!\\",\\"Der Pfeffer & Frost Geschenk Gutschein zum Einlösen im Shop - mache anderen eine
Freude

,

Bitte beachte: Es erfolgt KEIN Versand per Post!\\",,,abcgeschenkgutschein,10 EUR,10 EUR,new,,,,in stock,
PFE-Gutschein-20,Geschenkgutscheine 10€ - 100€ (20€),\\"Der Pfeffer & Frost Geschenk Gutschein zum Einlösen im Shop - mache anderen eine
Freude

,

Bitte beachte: Es erfolgt KEIN Versand per Post!\\",\\"Der Pfeffer & Frost Geschenk Gutschein zum Einlösen im Shop - mache anderen eine
Freude

,

Bitte beachte: Es erfolgt KEIN Versand per Post!\\",,,abcgeschenkgutschein,20 EUR,20 EUR,new,,,,in stock,
PFE-Gutschein-30,Geschenkgutscheine 10€ - 100€ (30€),\\"Der Pfeffer & Frost Geschenk Gutschein zum Einlösen im Shop - mache anderen eine
Freude

,

Bitte beachte: Es erfolgt KEIN Versand per Post!\\",\\"Der Pfeffer & Frost Geschenk Gutschein zum Einlösen im Shop - mache anderen eine
Freude

,

Bitte beachte: Es erfolgt KEIN Versand per Post!\\",,,abcgeschenkgutschein,30 EUR,30 EUR,new,,,,in stock,
PFE-Gutschein-50,Geschenkgutscheine 10€ - 100€ (50€),\\"Der Pfeffer & Frost Geschenk Gutschein zum Einlösen im Shop - mache anderen eine
Freude

,

Bitte beachte: Es erfolgt KEIN Versand per Post!\\",\\"Der Pfeffer & Frost Geschenk Gutschein zum Einlösen im Shop - mache anderen eine
Freude

,

Bitte beachte: Es erfolgt KEIN Versand per Post!\\",,,abcgeschenkgutschein,50 EUR,50 EUR,new,,,,in stock,
PFE-Gutschein-100,Geschenkgutscheine 10€ - 100€ (100€),\\"Der Pfeffer & Frost Geschenk Gutschein zum Einlösen im Shop - mache anderen eine
Freude

,

Bitte beachte: Es erfolgt KEIN Versand per Post!\\",\\"Der Pfeffer & Frost Geschenk Gutschein zum Einlösen im Shop - mache anderen eine
Freude

,

Bitte beachte: Es erfolgt KEIN Versand per Post!\\",,,abcgeschenkgutschein,100 EUR,100 EUR,new,,,,in stock,
buch-nudeln,asf,kkk,kkk,https://s3.triebwork.com/static/products/pfefferundfrost-nudelnmachengluecklich_kochbuch_katharinapflug_01.jpg,https://s3.triebwork.com/static/products/pfefferundfrost-nudelnmachengluecklich_kochbuch_katharinapflug_04.jpg,abcnudeln-machen-gluecklich,13.9 EUR,13.9 EUR,new,1234567891234,Pfeffer & Frost,,in stock,
"
`;
